File: src/main/java/com/harleylizard/magic_things/common/block/FouledSaplingBlock.java
package com.harleylizard.magic_things.common.block;

import com.harleylizard.magic_things.common.MagicThingsBlockTags;
import net.minecraft.core.BlockPos;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.block.SaplingBlock;
import net.minecraft.world.level.block.grower.TreeGrower;
import net.minecraft.world.level.block.state.BlockState;
import org.jetbrains.annotations.NotNull;

public final class FouledSaplingBlock extends SaplingBlock {

    public FouledSaplingBlock(TreeGrower grower, Properties properties) {
        super(grower, properties);
    }

    @Override
    protected boolean mayPlaceOn(@NotNull BlockState blockState, @NotNull BlockGetter blockGetter, @NotNull BlockPos blockPos) {
        return blockState.is(MagicThingsBlockTags.FOULED_SAPLING_GROWS_ON);
    }

}

File: src/main/kotlin/com/harleylizard/magic_things/common/block/FouledSaplingBlock.kt
package com.harleylizard.magic_things.common.block

import net.minecraft.core.BlockPos
import net.minecraft.world.level.BlockGetter
import net.minecraft.world.level.block.SaplingBlock
import net.minecraft.world.level.block.grower.TreeGrower
import net.minecraft.world.level.block.state.BlockState
import java.util.*

class FouledSaplingBlock(properties: Properties) : SaplingBlock(TreeGrower("fouled", 0.1f,
    Optional.empty(),
    Optional.empty(),
    Optional.of(MagicThingsConfiguredFeatures.fouledTree),
    Optional.empty(),
    Optional.empty(),
    Optional.empty()
), properties) {

    override fun mayPlaceOn(blockState: BlockState, blockGetter: BlockGetter, blockPos: BlockPos): Boolean {
        return blockState.`is`(MagicThingsBlockTags.fouledSaplingGrowsOn)
    }

}